library-best-practices/JLBP-3.md

```diff
@@ -0,0 +1,45 @@
+[JLBP-3] Use Semantic versioning
+--------------------------------
+
+Definition of semver: https://semver.org
+
+- For the purpose of compliance with semver, the stable surface of a library
+  serves as its "Public API." Annotations can be used to mark features
+  (classes, methods, etc) as unstable, and thus not part of the Public API,
+  such that semver rules don't have to apply. 
+  - Examples of annotations:
+    - Guava uses `@Beta`
+    - grpc-java uses `@ExperimentalApi`
+    - Google-cloud-java uses `@BetaApi`, `@InternalApi`, and
+      `@InternalExtensionOnly`
+  - Library documentation should let users know how to detect usage of
+    features that use these annotations
+  - Even though semver rules don't apply to unstable features, it is
+    recommended to bump a library's minor version if unstable features have
+    surface breakages.
+- Tools are available to help identify accidental incompatibilities within a
+  major version. Examples:
+  - https://lvc.github.io/japi-compliance-checker/ 
+  - http://www.mojohaus.org/clirr-maven-plugin/
+  - Example compatibility report for grpc-core:
+    https://abi-laboratory.pro/index.php?view=timeline&lang=java&l=grpc-core 
+- Examples of breaking changes that require a new major version:
```

Comment:
We need to call out 0.x versions here to indicate that this only applies when the version is >= 1.0. Per spec "Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable."

This goes hand in hand with a requirement to use and ship 1.0+ versions.

Reply:
Why do we need to call it out when it's already in the semver spec? Do you think people will be confused as this is now?

Reply:
Yes, as written I'd read this that 0.1 needs to go 1.0 when a breaking change is pushed.

Reply:
Ok, I will try to make that clearer.